File: yolo-nas-py/yolo_nas/draw.py
import numpy as np
import cv2


def draw_box(source, box, label, score, color, alpha=0.25):
    """Draw boxes on images"""
    # fill box
    crop_box = source[
        box[1] : (box[1] + box[3]), box[0] : (box[0] + box[2])
    ]  # crop box from source
    color_box = np.ones([*crop_box.shape[:2], 1], dtype=np.uint8) * np.asarray(
        color, dtype=np.uint8
    )  # color box (same size with crop). [h, w, 1] * [c] => [h, w, c]
    cv2.addWeighted(
        crop_box, 1 - alpha, color_box, alpha, 1.0, crop_box
    )  # weighted from color box to source

    cv2.rectangle(source, box, color, 2)  # draw box

    # measuring text
    size = min(source.shape[:2]) * 0.0007
    thickness = int(min(source.shape[:2]) * 0.001)
    (label_width, label_height), _ = cv2.getTextSize(
        f"{label} - {round(score, 2)}%",
        cv2.FONT_HERSHEY_SIMPLEX,
        size,
        thickness,
    )
    # draw labels (filled rect with text inside)
    cv2.rectangle(
        source,
        (box[0] - 1, box[1] - int(label_height * 2)),
        (box[0] + int(label_width * 1.1), box[1]),
        color,
        cv2.FILLED,
    )
    cv2.putText(
        source,
        f"{label} - {round(score, 2)}%",
        (box[0], box[1] - int(label_height * 0.7)),
        cv2.FONT_HERSHEY_SIMPLEX,
        size,
        [255, 255, 255],
        thickness,
        cv2.LINE_AA,
    )

File: yolo-nas-cpp/src/draw.cpp
#include <opencv2/opencv.hpp>
#include <opencv2/imgproc.hpp>

void check_boxes(cv::Mat &source, cv::Rect &box)
{
    if (box.x < 0)
        box.x = 0;
    if (box.y < 0)
        box.y = 0;
    if (box.x + box.width > source.cols)
        box.width = source.cols - box.x;
    if (box.y + box.height > source.rows)
        box.height = source.rows - box.y;
}

void draw_box(cv::Mat &source, cv::Rect &box, std::string &label, float &score, cv::Scalar color, float alpha)
{
    check_boxes(source, box); // check overflow boxes

    // Fill box
    cv::Mat cropBox = source(box);
    cv::Mat colorBox = cv::Mat::ones(cropBox.rows, cropBox.cols, CV_8UC3);
    colorBox = colorBox.setTo(color);
    cv::addWeighted(cropBox, 1 - alpha, colorBox, alpha, 1.0, cropBox);
    cv::rectangle(source, box, color, 2);
    cropBox.release();
    colorBox.release();

    double size = std::min<int>(source.cols, source.rows) * 0.0007;
    int thickness = (int)std::floor((float)std::min<int>(source.cols, source.rows) * 0.001f);
    int baseline = 0;
    std::string text = label + " - " + std::to_string(score * 100).substr(0, 5) + "%";
    cv::Size labelSize = cv::getTextSize(text, cv::FONT_HERSHEY_SIMPLEX, size, thickness, &baseline);
    baseline += thickness;

    cv::Rect textBox((int)(box.x - size), box.y - labelSize.height * 2,
                     (int)(labelSize.width * 1.05), labelSize.height * 2);
    cv::rectangle(source, textBox, color, cv::FILLED);
    cv::putText(source, text, cv::Point(box.x + 1, box.y - (int)(labelSize.height * 0.6)),
                cv::FONT_HERSHEY_SIMPLEX, size, cv::Scalar(255, 255, 255), thickness,
                cv::LINE_AA);
}